bugsnag-android-core/src/main/java/com/bugsnag/android/internal/DocumentPath.kt

```diff
@@ -0,0 +1,209 @@
+    private fun fillInMissingContainers(parent: Any?, index: Int): Any? {
+        var updatedParent = parent
+        if (index >= directives.size) {
+            return null
+        }
+        val directive = directives[index]
+        if (updatedParent == null) {
+            updatedParent = directive.newContainer()
+        }
+        val result = fillInMissingContainers(directive.getFromContainer(updatedParent), index + 1)
+        if (result != null) {
+            directive.setInContainer(updatedParent, result)
+        }
+        return updatedParent
+    }
+
+    // Navigate down into a container, then set the specified value.
+    private fun applyValueToContainer(value: Any?, container: Any) {
+        // These directives navigate to where we want to edit the document.
+        var updatedContainer = container
+        for (i in 0 until directives.size - 1) {
+            val directive = directives[i]
+            updatedContainer = directive.getFromContainer(updatedContainer)!!
+        }
+
+        // The last directive does the actual setting.
+        val directive = directives[directives.size - 1]
+        directive.setInContainer(updatedContainer, value)
+    }
+
+    companion object {
+        fun toPathDirectives(path: String): List<DocumentPathDirective> {
+            val directives: MutableList<DocumentPathDirective> = ArrayList(0)
+            if (path.isEmpty()) {
+                return directives
+            }
+            var requiresEscaping = false
+            var strBegin = 0
+            var i = 0
+            while (i < path.length) {
+                when (path[i]) {
+                    '\\' -> {
+                        requiresEscaping = true
+                        i++
+                    }
+                    '.' -> {
+                        val pathComponent: String = if (requiresEscaping) {
+                            unescape(path, strBegin, i)
+                        } else {
+                            path.substring(strBegin, i)
+                        }
+                        directives.add(makeDirectiveFromPathComponent(pathComponent))
+                        strBegin = i + 1
+                        requiresEscaping = false
+                    }
+                    else -> {
+                    }
+                }
+                i++
+            }
+            if (strBegin < path.length) {
+                val pathComponent: String = if (requiresEscaping) {
+                    unescape(path, strBegin, path.length)
+                } else {
+                    path.substring(strBegin)
+                }
+                directives.add(makeDirectiveFromPathComponent(pathComponent))
+            }
+            if (isDotLast(path)) {
+                directives.add(DocumentPathDirective.ListInsertDirective())
+            }
+            return directives
+        }
+
+        private fun unescape(str: String, startOffset: Int, endOffset: Int): String {
```

Comment:
`String.replace()` is very slow...

Reply:
Has that been benchmarked? If this has been shown to be faster then that's fine as long as there's a comment explaining it's an optimization, otherwise IMO it'd be best to stick with `String.replace()` as it's the least surprising choice.

Reply:
Yes it came up during my optimization phase as I was trying to get the string processing time down. Basically anything that calls into the regex system suffers a massive slowdown (like 2x, 3x in some cases) vs doing it by hand. I'm adding comments to explain the choices.